client/lib/explat/internals/anon-id.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * Internal dependencies
+ */
+import { recordTracksEvent } from 'calypso/lib/analytics/tracks';
+import { tracksAnonymousUserId } from 'calypso/lib/analytics/ad-tracking';
+import userUtils from 'calypso/lib/user/utils';
+import { logError, isDevelopmentMode } from './log-error';
+
+// Using typescript to ensure we are being safe in SSR
+declare const window: undefined | ( Window & typeof globalThis );
+
+const logErrorOrThrowInDevelopmentMode = ( message: string ) => {
+	if ( isDevelopmentMode ) {
+		throw new Error( message );
+	} else {
+		logError( { message } );
+	}
+};
+
+/**
+ * setInterval, but it runs first callback immediately instead of after interval.
+ *
+ * @param f The callback function
+ * @param intervalMs The interval in milliseconds
+ */
+const immediateStartSetInterval = ( f: () => void, intervalMs: number ) => {
+	f();
+	return setInterval( f, intervalMs );
+};
+
+let initializeAnonIdPromise: null | Promise< string | null > = null;
+const anonIdPollingIntervalMilliseconds = 50;
+const anonIdPollingIntervalMaxAttempts = 100; // 50 * 100 = 5000 = 5 seconds
+/**
+ * Initialized the anonId
+ * - Polls for AnonId receival
+ * - Should only be called once at startup
+ * - Happens to be safe to call multiple times if it is necessary to reset the anonId - something like this was necessary for testing.
+ *
+ * The state of initializeAnonIdPromise should be used rather than the return of this function.
+ * The return is avaliable to make this easier to test.
+ *
+ * Throws on error.
+ *
+ * TODO: More error checking?
+ * TODO: Better timing?
+ * TODO: Maybe look further into the tracks event system to see if there is an alternative method for this.
```

Comment:
Yeah, I'm not sure that this code is required in its current form, and if it is, it shouldn't be an ExPlat-only thing. 

Looking at the `calypso-analytics` package, it exposes [`initializeAnalytics()`](https://github.com/Automattic/wp-calypso/blob/da5b69a6ed89e7494bf468b0ff25ff0763e49376/packages/calypso-analytics/src/tracks.ts#L131), which seems to guarantee that the `tk_ai` cookie is set if needed (either by the server or by a client-side fallback).

I'm not sure if it's safe to call `initializeAnalytics()` multiple times. If it isn't, we can fix it (or add a wrapper) to the `calypso-analytics` package and call that, obviating the need for this function.

Reply:
Very good point! I didn't look deeply enough into the tracks code and should have at least asked about this 🤦

Reply:
It seems like `initializeAnalytics()` sets up tracks but it does not initialise the `anonId`, for that it needs to either send an initial tracks event - which it doesn't do unless there is given user data AFAICT, or send an [`identifyAnonUser` tracks command](https://github.com/Automattic/analytics/blob/32a95102569ec42ef1643553626e87897a82f796/lib/w/tracks.js#L322).
 
What might be misleading is that it returns a promise that comes from [`checkForBlockedTracks()`](https://github.com/Automattic/wp-calypso/blob/3c6f806059255c46cdfe58bba841dfb652690302/packages/calypso-analytics/src/tracks.ts#L78-L109), this doesn't check that tracks events have been sent but rather whether tracks has been initialised - e.g. blocked by an add blocker as `w2.js` isn't yet bundled with Calypso.